File: Unity/RLBridge/Scripts/Messages.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace RLBridge
{

    public abstract class MsgObj
    {
        public string toJson()
        {
            return JsonUtility.ToJson(this);
        }
    }

    public class Message : MsgObj
    {
        public string title;
        public string json_content;

        public Message(string title, string json_content)
        {
            this.title = title;
            this.json_content = json_content;
        }

    }

    public class BeginConnectionRequestMsg : MsgObj
    {
        public string behavior_name;

        public BeginConnectionRequestMsg(string behavior_name)
        {
            this.behavior_name = behavior_name;
        }
    }

    public class BeginConnectResponseMsg : MsgObj
    {
        public int observation_size;
        public int continuous_action_size;
        public int discrete_action_size;
        public string[] agents;

        public BeginConnectResponseMsg(int observation_size, int continuous_action_size, int discrete_action_size, string[] agents)
        {
            this.observation_size = observation_size;
            this.continuous_action_size = continuous_action_size;
            this.discrete_action_size = discrete_action_size;
            this.agents = agents;
        }
    }

    public class ActionMsg : MsgObj
    {
        public string behavior_name;
        public string agent_name;
        public double[] continuous_actions;
        public int[] discrete_actions;

        public ActionMsg(string behavior_name, string agent_name, double[] continuous_actions, int[] discrete_actions)
        {
            this.behavior_name = behavior_name;
            this.agent_name = agent_name;
            this.continuous_actions = continuous_actions;
            this.discrete_actions = discrete_actions;
        }
    }

    public class StatusMsg : MsgObj
    {
        public string behavior_name;
        public string agent_name;
        public double[] observations;
        public float reward;
        public bool done;

        public StatusMsg(string behavior_name, string agent_name, double[] observations, float reward, bool done)
        {
            this.behavior_name = behavior_name;
            this.agent_name = agent_name;
            this.observations = observations;
            this.reward = reward;
            this.done = done;
        }

    }

    public class ResetMsg : MsgObj
    {
        public string behavior_name;
        public string agent_name;

        public ResetMsg(string behavior_name, string agent_name)
        {
            this.behavior_name = behavior_name;
            this.agent_name = agent_name;
        }
    }
}

File: Python/RLBridge/messages.py
import json

class MsgObj:
    def toJson(self):
        return json.dumps(self.__dict__)

class Message(MsgObj):

    def __init__(self,title,json_content):
        self.title = title
        self.json_content=json_content

    
class ActionMsg(MsgObj):

    def __init__(self,behavior_name,agent_name,continuous_actions,dicrete_actions):
        self.behavior_name = behavior_name
        self.agent_name = agent_name
        self.continuous_actions = continuous_actions
        self.discrete_actions = dicrete_actions
    
class StatusMsg(MsgObj):

    def __init__(self,behavior_name,agent_name,observations,reward,done):
        self.behavior_name = behavior_name
        self.agent_name = agent_name
        self.observations = observations
        self.reward = reward
        self.done = done

class ResetMsg(MsgObj):

    def __init__(self,behavior_name,agent_name):
        self.behavior_name = behavior_name
        self.agent_name = agent_name

class BeginConnectionRequestMsg(MsgObj):

    def __init__(self,behavior_name):
        self.behavior_name = behavior_name

class BeginConnectionResponseMsg(MsgObj):

    def __init__(self,observation_size,continuous_action_size,discrete_action_size,agent_names):
        self.observation_size = observation_size
        self.continuous_action_size = continuous_action_size
        self.discrete_action_size = discrete_action_size
        self.agent_names = agent_names
    
def JsonToMessage(json_content):
    return Message(**json.loads(json_content))

def JsonToActionMsg(json_content):
    return ActionMsg(**json.loads(json_content))

def JsonToStatusMsg(json_content):
    return StatusMsg(**json.loads(json_content))

def JsonToResetMsg(json_content):
    return ResetMsg(**json.loads(json_content))

def JsonToBeginConnectionResponseMsg(json_content):
    return BeginConnectionResponseMsg(**json.loads(json_content))

def JsonToBeginConnectionRequestMsg(json_content):
    return BeginConnectionRequestMsg(**json.loads(json_content))
    
